src/metabase/notification/payload/core.clj

```diff
@@ -0,0 +1,75 @@
+(ns metabase.notification.payload.core
+  (:require
+   [metabase.email.messages :as messages]
+   [metabase.models.notification :as models.notification]
+   [metabase.public-settings :as public-settings]
+   [metabase.pulse.core :as pulse]
+   [metabase.util :as u]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.schema :as ms]))
+
+(def Notification
+  "Schema for the notification."
+  ;; TODO: how do we make this schema closed after :merge?
+  [:merge #_{:closed true}
+   [:map
+    [:payload_type                  (into [:enum] models.notification/notification-types)]
+    [:id           {:optional true} ms/PositiveInt]
+    [:active       {:optional true} :boolean]
+    [:created_at   {:optional true} :any]
+    [:updated_at   {:optional true} :any]]
+   [:multi {:dispatch :payload_type}
+    ;; system event is a bit special in that part of the payload comes from the event itself
+    [:notification/system-event [:map
+                                 [:payload
+                                  [:map {:closed true}
+                                   ;; TODO: event-info schema for each event type
+                                   [:event-topic [:fn #(= "event" (-> % keyword namespace))]]
+                                   [:event-info  [:maybe :map]]]]]]
+
+    ;; for testing only
+    [:notification/testing       :map]]])
+
+(def NotificationPayload
+  "Schema for the notification payload."
+  ;; TODO: how do we make this schema closed after :merge?
+  [:merge
+   Notification
+   [:map
+    [:context [:map]]]
+   [:multi {:dispatch :payload_type}
+    ;; override payload to add extra-context key
+    [:notification/system-event [:map
+                                 ;; override the payload with extra context
+                                 [:payload
+                                  [:map {:closed true}
+                                   [:event-topic                    [:fn #(= "event" (-> % keyword namespace))]]
+                                   [:event-info                     [:maybe :map]]
+                                   [:extra-context {:optional true} [:maybe :map]]]]]]
```

Comment:
I'm wondering if it makes sense to have `payload.extra` kind of like `context.extra`, so that template authors don't need to remember which is which.